tests/mail/tests.py

```diff
@@ -92,6 +92,37 @@ def test_ascii(self):
         self.assertEqual(message["From"], "from@example.com")
         self.assertEqual(message["To"], "to@example.com")
 
+    @mock.patch("django.core.mail.message.MIMEText.set_payload")
+    def test_nonascii_as_string_with_ascii_charset(self, mock_set_payload):
+        """Line length check should encode the payload supporting `surrogateescape`.
+
+        Following https://github.com/python/cpython/issues/76511, newer
+        versions of Python (3.11.9, 3.12.3 and 3.13) ensure that a message's
+        payload is encoded with the provided charset and `surrogateescape` is
+        used as the error handling strategy.
+
+        This test is heavily based on the test from the fix for the bug above.
+        Line length checks in SafeMIMEText's set_payload should also use the
+        same error handling strategy to avoid errors such as:
+
+        UnicodeEncodeError: 'utf-8' codec can't encode <...>: surrogates not allowed
+
+        """
+
+        def simplified_set_payload(instance, payload, charset):
+            instance._payload = payload
+
+        mock_set_payload.side_effect = simplified_set_payload
+
+        text = (
+            "Text heavily based in Python's text for non-ascii messages: Föö bär"
+        ).encode("iso-8859-1")
+        body = text.decode("ascii", errors="surrogateescape")
+        email = EmailMessage("Subject", body, "from@example.com", ["to@example.com"])
+        message = email.message()
+        mock_set_payload.assert_called_once()
+        self.assertEqual(message.get_payload(decode=True), text)
```

Comment:
Do we need to copy tests from Python repo? We already have a test that crashes on the latest Python versions. I don't see much value in doing this.

Reply:
Our Jenkins Python 3.11 tests are using Python 3.11.4 and so [django-4.2](https://djangoci.com/job/django-4.2/) and [django-5.0](https://djangoci.com/job/django-5.0/) are still green without the backport. Ideally, we would update the CI to test Python 3.11.9+ and then `MailTests.test_dont_base64_encode` would start failing. That didn't look trivial and so this test aimed to cover it from a different angle.

Reply:
>  Ideally, we would update the CI to test Python 3.11.9+ and then `MailTests.test_dont_base64_encode` would start failing.

np, will do this later today.

Reply:
Thank you ⭐

Reply:
Thank you both for the comments!

While it's true that a test fails in newer Python versions if we remove the `surrogateescape` errors, I believe it's still important to maintain coverage for the original fix across all Python versions, including those where the failure doesn't occur.

My rationale is that the fix impacts a piece of code that is not only used in newer Python versions but is also employed in every version that our code supports. Therefore, I consider the test to be a valuable addition. Additionally, it's worth noting that the test is not directly copied from Python's codebase; instead, it's based on it but adapted to fit the context of Django. Specifically, it targets our custom code to check for the maximum line length in each raw email line.